native/java/org/jpype/JPypeContext.java

```diff
@@ -128,75 +130,98 @@ public void run()
    * fashion. Inherently this is a very dangerous time as portions of Java have
    * already been deactivated.
    */
-  private void watchShutdown()
+  @SuppressWarnings(
+          {
+            "CallToThreadYield", "SleepWhileInLoop"
+          })
+  private void shutdown()
   {
-    // Try to yield in case there is a race condition.  The user
-    // may have installed a shutdown hook, but we cannot verify
-    // the order that shutdown hook threads are executed.  Thus we will
-    // try to intentionally lose the race.
-    //
-    // This will only occur if something registered a shutdown hook through
-    // a Java API.  Those registered though the JPype API will be joined
-    // manually.
-    for (int i = 0; i < 5; i++)
+    try
     {
-      try
+      // Try to yield in case there is a race condition.  The user
+      // may have installed a shutdown hook, but we cannot verify
+      // the order that shutdown hook threads are executed.  Thus we will
+      // try to intentionally lose the race.
+      //
+      // This will only occur if something registered a shutdown hook through
+      // a Java API.  Those registered though the JPype API will be joined
+      // manually.
+      for (int i = 0; i < 5; i++)
       {
-        Thread.sleep(1);
-        Thread.yield();
-      } catch (InterruptedException ex)
+        try
+        {
+          Thread.sleep(1);
+          Thread.yield();
+        } catch (InterruptedException ex)
+        {
+        }
+      }
+
+      // Execute any used defined shutdown hooks registered with JPype.
+      if (!this.shutdownHooks.isEmpty())
       {
+        for (Thread thread : this.shutdownHooks)
+        {
+          thread.start();
+        }
+        for (Thread thread : this.shutdownHooks)
+        {
+          try
+          {
+            thread.join();
+          } catch (InterruptedException ex)
+          {
+          }
+        }
       }
-    }
 
-    // Execute any used defined shutdown hooks registered with JPype.
-    if (!this.shutdownHooks.isEmpty())
-    {
-      for (Thread thread : this.shutdownHooks)
+      // Disable all future calls to proxies
+      this.shutdownFlag.incrementAndGet();
+
+      // Past this point any further execution of a Python proxy would
+      // be fatal.
+      Thread t1 = Thread.currentThread();
+      Map<Thread, StackTraceElement[]> threads = Thread.getAllStackTraces();
+
+      for (Thread t : threads.keySet())
       {
-        thread.start();
+        if (t1 == t)
+          continue;
+//      if (t.getState() == Thread.State.RUNNABLE)
+        t.interrupt();
```

Comment:
Everyone gets an interrupt so that nothing is left waiting.  This doesn't hit Python unless it was waiting on a blocking Java call.